holoviews/element/raster.py

```diff
@@ -616,163 +632,63 @@ class QuadMesh(Raster):
 
+        Note: Deprecate as part of 2.0
+        """
+        if 'interface' not in state:
+            self.interface = GridInterface
+            x, y = state['_kdims_param_value']
+            z = state['_vdims_param_value'][0]
+            data = state['data']
+            state['data'] = {x.name: data[0], y.name: data[1], z.name: data[2]}
+        super(Dataset, self).__setstate__(state)
 
-    @classmethod
-    def collapse_data(cls, data_list, function, kdims=None, **kwargs):
+    def trimesh(self):
         """
-        Allows collapsing the data of a number of QuadMesh
-        Elements with a function.
+        Converts a QuadMesh into a TriMesh.
         """
-        if not all(data[0].ndim == 1 for data in data_list):
-            raise Exception("Collapsing of non-grid based QuadMesh"
-                            "currently not supported")
-        xs, ys, zs = zip(data_list)
-        if isinstance(function, np.ufunc):
-            z = function.reduce(zs)
-        else:
-            z = function(np.dstack(zs), axis=-1, **kwargs)
-        return xs[0], ys[0], z
-
-
-    def _coord2matrix(self, coord):
-        return tuple((np.digitize([coord[i]], self.data[i])-1)[0]
-                     for i in [1, 0])
-
-
-    def range(self, dimension, data_range=True):
-        idx = self.get_dimension_index(dimension)
-        dim = self.get_dimension(dimension)
-        if idx in [0, 1, 2] and data_range:
-            data = self.data[idx]
-            lower, upper = np.nanmin(data), np.nanmax(data)
-            return dimension_range(lower, upper, dim)
-        return super(QuadMesh, self).range(dimension, data_range)
-
-
-    def dimension_values(self, dimension, expanded=True, flat=True):
-        idx = self.get_dimension_index(dimension)
-        data = self.data[idx]
-        if idx in [0, 1]:
-            # Handle grid
-            if not self._grid:
-                return data.flatten()
-            odim = self.data[2].shape[idx] if expanded else 1
-            vals = np.tile(np.convolve(data, np.ones((2,))/2, mode='valid'), odim)
-            if idx:
-                return np.sort(vals)
-            else:
-                return vals
-        elif idx == 2:
-            # Value dimension
-            return data.flatten() if flat else data
-        else:
-            # Handle constant dimensions
-            return super(QuadMesh, self).dimension_values(idx)
+        xs = self.interface.coords(self, 'x', edges=True)
+        ys = self.interface.coords(self, 'y', edges=True)
+        if xs.ndim == 1:
+            xs, ys = (np.tile(xs[:, np.newaxis], len(ys)).T,
+                      np.tile(ys[:, np.newaxis], len(xs)))
+        zarr = self.dimension_values(2, flat=False)
+        s0 = zarr.shape[0]
+
+        t1 = np.arange(len(self))
+        js = (t1//s0)
+        t1s = js*(s0+1)+t1%s0
+        t2s = t1s+1
+        t3s = (js+1)*(s0+1)+t1%s0
+        t4s = t2s
+        t5s = t3s
+        t6s = t3s+1
```

Comment:
This is pretty unreadable and the variable names are uninformative.

Reply:
If this is doing some self-contained computation, it would be better as a utility which can at least have a docstring explaining what this is supposed to be doing.

Reply:
Sure, a utility is fine.

Reply:
End up with all kinds of circular import issues if I try to move it to ``element.util``, I've added various comments which should clarify.